src/likelihoods/negativebinomial.jl

```diff
@@ -1,36 +1,109 @@
+Each `NBParam` has its own documentation:
+- [`NBParamI`](@ref): This is the definition used by Distributions.jl.
+- [`NBParamII`](@ref): This is the definition used by Wikipedia.
+- [`NBParamIII`](@ref): This is the definition based on the mean.
 
-```math
-    p(k|successes, f) = \\frac{\\Gamma(k+successes)}{k! \\Gamma(successes)} l(f)^successes (1 - l(f))^k
-```
-On calling, this returns a negative binomial distribution with `successes` successes and 
-probability of success equal to `l(f)`.
-
-!!! warning "Parameterization" 
-    The parameter `successes` is different from the parameter `r` in the 
-    [Wikipedia definition](http://en.wikipedia.org/wiki/Negative_binomial_distribution), 
-    which denotes the number of failures.
-    This parametrization is used in order to stay consistent with the parametrization in 
-    [Distributions.jl](https://juliastats.org/Distributions.jl/stable/univariate/#Distributions.NegativeBinomial).
-    To use the Wikipedia definition, set `successes` as the number of "failures" and
-    change the probability of success from `l(f)` to `1 - l(f)`.
-    Note that with symmetric functions like the [`LogisticLink`](@ref), this corresponds to
-    using `l(-f)`.
-"""
-struct NegativeBinomialLikelihood{Tl<:AbstractLink,T<:Real} <: AbstractLikelihood
-    successes::T    # number of successes parameter
+To create a new parametrization, you need to;
+- Create a new typee from `struct MyNBParam <: NBParam end`
+- Dispatch [`to_dist_params`](@ref): `to_dist_params(l::NegativeBinomialLikelihood{MyNBParam}, f::Real)`
+to return the parameters for the [`NegativeBinomial`](https://juliastats.org/Distributions.jl/latest/univariate/#Distributions.NegativeBinomial) constructor
+- Write a constructor for `NegativeBinomialLikelihood{MyNBParam}`
+"""
+struct NegativeBinomialLikelihood{NBParam,Tl<:AbstractLink,T} <: AbstractLikelihood
+    params::T # Likelihood parameters (depends of NBParam)
     invlink::Tl
+    function NegativeBinomialLikelihood{Tparam}(params, invlink) where {Tparam}
+        invlink = link(invlink)
+        return new{Tparam,typeof(invlink),typeof(params)}(params, invlink)
+    end
 end
 
-function NegativeBinomialLikelihood(l=logistic; successes::Real=1)
-    return NegativeBinomialLikelihood(successes, link(l))
-end
+@deprecate NegativeBinomialLikelihood(l=logistic; successes=1) NegativeBinomialLikelihood{
+    NBParamI
+}(
+    l; successes
+)
 
 @functor NegativeBinomialLikelihood
 
-(l::NegativeBinomialLikelihood)(f::Real) = NegativeBinomial(l.successes, l.invlink(f))
+function (l::NegativeBinomialLikelihood)(f::Real)
+    return NegativeBinomial(to_dist_params(l, f)...)
+end
 
 (l::NegativeBinomialLikelihood)(fs::AbstractVector{<:Real}) = Product(map(l, fs))
+
+"""
+    to_dist_params(l::NegativeBinomialLikelihood{NBParam}, f::Real)->Tuple{Real,Real}
+
+Take parameters from a given parametrization `NBParam` to the parametrization
+used by `Distributions.jl`
+"""
+to_dist_params
```

Comment:
Wouldn't it be easier to just require that for new parameterizations `(l::NegativeBinomialLikelihood{MyParam})(f::Real)` is implemented? This would avoid adding anything new to the API that users/developers have to take care of and know about.

Reply:
Yeah, you're probably right